### PR TITLE
Performance improvements with large strings

### PR DESCRIPTION
**Pull request by leohemsted**
TL;DR: Speed up `_tokenize` by using a 3rd party regex lib.

# The problem: 

`_tokenize` identifies html tags in text (since smartypants doesn't touch stuff inside tags). Smartypantsifying in a large (8400 character) string with no tags or even angle brackets in was proving to be prohibitively slow (0.5 seconds).

# The solution:

`regex` is a [PCRE](https://www.pcre.org/)-compliant 3rd party library - we can take advantage of [atomic groups](https://www.regular-expressions.info/atomic.html) to prevent backtracking attempts (the second half of the regex, which matches the tag, would try to backtrack into the first group, of non `<` characters, trying to find a `<`). 

This is an optional improvement - if `regex` is not installed it'll fall back to the older, slower regex.

Speed up is significant - from 0.58 to 0.037 for my test input.

image

With thanks to James Kirrage

### CODE REVIEW

File: smartypants.py [hunk 1]
```diff
@@ -563,11 +569,15 @@ def _tokenize(text):
...
+    If you have the ``regex`` library (https://pypi.python.org/pypi/regex/),
+    this function will use an alternative regex to perform significantly faster
+    on large input texts.
+    """
     tokens = []
 
-    tag_soup = re.compile(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', re.S)
+    tag_soup = re.compile(TAG_REGEX, re.S)
 
     token_match = tag_soup.search(text)
```
Comment (allait) [inline smartypants.py]: As far as I understand, this is the part where backtracking happens. It's not the pattern itself that goes back, it's the `.search`, which makes python retry the `[^<]*`match starting from each of the string characters:

But since our pattern starts with a wildcard match and we rerun it after matching a tag from the end of previous match I think we can replace this (and the repeated call within the loop) with `.match`, which will only try to apply the pattern from the beginning of the string. But I might be missing something about why we need `.search` there.

```
In [24]: %timeit re.search(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', 'hello'*2000)
1.02 s ± 71.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [25]: %timeit re.match(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', 'hello'*2000)
202 µs ± 4.92 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

# This doesn't happen when there is a match

In [26]: %timeit re.match(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', 'hello<tag>'*2000)
2.34 µs ± 214 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [27]: %timeit re.search(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', 'hello<tag>'*2000)
2.17 µs ± 113 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

**Comment by leohemsted** [comment]
deprecated in favour of https://github.com/leohemsted/smartypants.py/pull/2
